src/CodeGen_X86.cpp

```diff
@@ -509,7 +517,9 @@ void CodeGen_X86::codegen_vector_reduce(const VectorReduce *op, const Expr &init
 }
 
 string CodeGen_X86::mcpu() const {
-    if (target.has_feature(Target::AVX512_Cannonlake)) {
+    if (target.has_feature(Target::AVX512_SapphireRapids)) {
```

Comment:
Do the changes in this method also need `#if LLVM_VERSION >= 120`?

Reply:
My thinking was to restrict where the feature was set, rather than where the feature is used. I'm not sure which approach would require the least `#if`s or the best errors for users.

Currently these changes will prevent Halide automatically selecting SapphireRapids on LLVM 10 or 11, but will still cause problems if a user provides `HL_TARGET=x86-64-linux-avx512_sapphirerapids`. Would it make sense to add the `#if` to prevent parsing this string for LLVM <12?

Reply:
I guess it depends on whether the changes you are making are entirely in Halide (in which case the LLVM version doesn't matter) or if they rely on LLVM codegen (in which case it does). Which is the case?

Reply:
I'm not sure I understand the question. The changes should involve only changing Halide code (not LLVM code) but will rely heavily on the intrinsics being available in LLVM, which they are for 12+.

I have changed this to copy what is being done for CUDA 80. The automatic feature selection is guarded by `LLVM_VERSION`, but if a user deliberately requests it then it will raise an error in `CodeGen_X86::mcpu` and `CodeGen_X86::mattr` for LLVM < 12.